### PR TITLE
Metric Learning of Melanoma Classification

### PR DESCRIPTION
**Pull request by tranjamin**
This pull request is to classify benign and malignant skin lesions using a siamese network. It uses a ResNet50 backbone to extract salient features from the images and create an embeddings space where the two classes are as far away from each other as possible. 

Author: Benjamin Tran
Student Number: 46969196

Files Modified:
- `train.py`: the file used to train the model
- `predict.py`: the file used to predict the model
- `util.py`: contains visualisation helper functions
- `mnist.py`: contains an example of using Siamese networks on MNIST
- `generate_balanced_dataset.py`: generates the smaller balanced dataset
- `modules.py`: contains network definitions for siamese networks
- `requirements.txt`: used to load in the required Python environment
- `dataset.py`: contains dataset loader functions
- `Modules/`: a reference to an external helper library.

(Note: the Blackboard submitted README is out of date)

### CODE REVIEW

**Comment by tranjamin** [comment]
Future work:
To transition from ResNet to MobileNet, which is showing indicators of better performance

**Comment by tranjamin** [comment]
This current iteration now utilises Inception Net with a better oversampling strategy

**Comment by tranjamin** [comment]
This iteration now has options to select different types of models and datasets. The default arguments used should be `--full-dataset`.

**Comment by hanemma7moud** [comment]
### This is an initial inspection, no action is required at this point
Recognition Problem : total : **19**

Solves problem: The solution is well-suited to the problem (**4**) but the validation accuracy and other paramaters havehigh fluctuations.
The training manifold is not well separated
 the classifier shows good performance in training but not the validation.
Triplet loss is utilized, which is suitable for the problem.
Class imbalance is addressed through oversampling the benign cases.

Implementation functions as intended (**2**)

Good design: Well-designed (**1**)

Commenting: Clear and sufficient comments throughout the code. (**1**)

Difficulty: Hard (**10**)

Note:

+Well understaning of the problem and the usgae of metric leaning to solve it along with its challanges.

- Validation Loss: Downward trend with high fluctuations; suggests instability. Try lowering learning rates or adding regularization.
- Validation Accuracy: Inconsistent, possibly not converged. Consider monitoring for more epochs and implementing early stopping.
- Precision & Recall: Significant variation, possibly due to class imbalance. Address with class weighting or data augmentation.
- Overall Stability: Add techniques like gradient clipping, batch normalization, or learning rate schedules to reduce oscillations.

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>No Header blocks. </td><td>-1</td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. The training manifold is not well separated.</td><td>-1</td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. Has data splits.</td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard.  Siamese network (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Acceptable. References missing.</td><td>-1</td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF Checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>-3</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_
